doc/developer/design/20230806_durable_catalog_state.md

```diff
@@ -0,0 +1,402 @@
+
+Once the migrations are de-tangled, at a high level they will take the following steps:
+
+1. Use `impl From<objects_v(X - 1)::(Object)> for objects_v(X)::(Object)` to convert every object
+   from their old format into their new format.
+2. Use `mz_adapter::catalog::storage::Transaction` to perform any deletions, insertions, or updates
+   that require read/write semantics instead of `mz_stash::transaction::Transaction`.
+
+As a bonus, these migrations start to look very similar to the
+existing [catalog migrations](https://github.com/MaterializeInc/materialize/blob/main/src/adapter/src/catalog/migrate.rs).
+We might be able to combine them into a single logical step instead of having two separate migration
+frameworks (not counting the builtin migration framework).
+
+## Alternatives
+
+- Use multiple shards for the persist implementation
+  once [multi-shard txns](https://github.com/MaterializeInc/materialize/pull/20954) is merged and
+  stable.
+- An alternative approach is to put a trait directly in front of the stash, and implement that trait
+  using the current CRDB implementation and implement the trait using persist.
+    - Pros:
+        - Re-using the existing stash migration might be easier.
+        - We can re-use the existing stash debug tool.
+        - The controller can use persist for durable state for free.
+    - Cons:
+        - Maintaining a generic durable TVC store adds a lot of complexity to the implementation.
+          Much of that complexity is already handled by persist, which is a generic durable TVC
+          store with a dedicated team. Overall, I think it might be a net-win if we can at some
+          point delete the stash crate because it removes a layer of complexity from the adapter.
+        - It's very difficult to get the stash trait right and maintain it. We previously had a
+          stash trait but removed it because it was too hard to maintain.
+- The persist implementation could not use an in-memory cache, and instead go to persist for every
+  read via a read handle. This has better memory usage, but slower reads. The reason that I decided
+  on the current proposed implementation is that the Coordinator already reads the entire durable
+  catalog state into memory for every catalog transaction. So we already need that amount of memory
+  free at any time in order to be able to serve DDL. Certain append only object types, like audit
+  logs and storage usage, don't need to be cached in memory.
+
+## Open questions
+
+- Should the `DurableCatalogTrait` be broken up into two traits? A read only trait and a read/write
+  trait?
+- For the persist implementation, can we use a schema like `tag: u64, raw: bytes` and take advantage
+  of persist filtering based on the tag?
+- Is checking that the persist shard upper sufficient for a leadership check or do we need to check
+  the epoch?
+- Should this work live in its own crate or stay part of the adapter crate? My opinion is that it
```

Comment:
own crate!

Reply:
A followup question to this that I started to ponder after writing this is, should we move the whole catalog or just the storage module?

Reply:
ditto own crate